### PR TITLE
HDDS-6481. Add metrics for S3 gateway

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

Currently S3 gateway don't have a standard metrics, this ticket is to add the metrics for S3 gateway to record the successful and failed operations.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6481

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by symious** [comment]
@adoroszlai @ferhui @bharatviswa504 Could you help to review this PR?

**Comment by kuenishi** [comment]
Hey, thanks for this work. I also wanted the feature for S3 API stats. But I also wanted JVM metrics as well as other services, is there any chance JVM metrics added into this pull request?

**Comment by symious** [comment]
@kuenishi Thanks for the advice.
Added JvmMetrics, please have a check.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+   * Unregister the metrics instance.
+   */
+  public static void unRegister() {
+    instance = null;
+    MetricsSystem ms = DefaultMetricsSystem.instance();
+    ms.unregisterSource(SOURCE_NAME);
+  }
+
+  @Override
+  @SuppressWarnings("SuspiciousMethodCalls")
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:119]: Is this `SuppressWarnings` necessary?

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    deleteKeyFailure.snapshot(recordBuilder, true);
+  }
+
+  // INC
+  public void incGetBucketSuccess() {
+    getBucketSuccess.incr();
+  }
+
+  public void incGetBucketFailure() {
+    getBucketSuccess.incr();
```
Comment (adoroszlai) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/metrics/S3GatewayMetrics.java:171]: Typo:

```suggestion
    getBucketFailure.incr();
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @symious for the patch.  Mostly looks good, except one typo and a question.  Also, can you please make Prometheus aware of this by adding `s3g:9878` here:

https://github.com/apache/ozone/blob/1baa6521d125fcdf03a0e65d85eadbc60e1a5437/hadoop-ozone/dist/src/main/compose/ozone/prometheus.yml#L23-L25

**Comment by symious** [comment]
@adoroszlai Updated according to the comments, please have a check :)

**Comment by adoroszlai** [comment]
Thanks @symious for updating the patch, LGTM.  Let's wait for others to give feedback, too.

**Review by ferhui — APPROVED** [review]
Thanks, LGTM

**Comment by adoroszlai** [comment]
@kuenishi would you like to take another look?

**Comment by adoroszlai** [comment]
Thanks @symious for the patch, @ferhui and @kuenishi for the review.

**Pull request merged**